### PR TITLE
Fix potential memory leak in ebpf.plugin

### LINKED ISSUE #9424: Convert calls from error to info inside ebpf.plugin
##### Bug report summary
As we discussed on https://github.com/netdata/netdata/issues/9390#issuecomment-649661450 some messages should be classified as info by eBPF. plugin

##### OS / Environment
All Linux

##### Netdata version
netdata 1.23
 

##### Component Name
Collector

##### Steps To Reproduce
1. Start Netdata 1.23.0
2. Read error.log

##### Expected behavior
Reduce the level of information for our users, because it is giving a wrong impression about events happened with the collector.

### LINKED ISSUE #9480: KernelCare and Netdata conflicts with kprobe events
##### Bug report summary

I installed Netdata on cPanel server with KernelCare. KernelCare is unable to apply kernel patches due to kprobe events created by the netdata.
We reached out kernel care support and they investigated the issue with Kernel Care Patch.  Find an image for reference below
http://prntscr.com/tbt6ih
##### OS / Environment

```
Linux node.server.com 3.10.0-1127.el7.x86_64 #1 SMP Tue Mar 31 23:36:51 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
/etc/centos-release:CentOS Linux release 7.8.2003 (Core)
/etc/os-release:NAME="CentOS Linux"
/etc/os-release:VERSION="7 (Core)"
/etc/os-release:ID="centos"
/etc/os-release:ID_LIKE="rhel fedora"
/etc/os-release:VERSION_ID="7"
/etc/os-release:PRETTY_NAME="CentOS Linux 7 (Core)"
/etc/os-release:ANSI_COLOR="0;31"
/etc/os-release:CPE_NAME="cpe:/o:centos:centos:7"
/etc/os-release:HOME_URL="https://www.centos.org/"
/etc/os-release:BUG_REPORT_URL="https://bugs.centos.org/"
/etc/os-release:
/etc/os-release:CENTOS_MANTISBT_PROJECT="CentOS-7"
/etc/os-release:CENTOS_MANTISBT_PROJECT_VERSION="7"
/etc/os-release:REDHAT_SUPPORT_PRODUCT="centos"
/etc/os-release:REDHAT_SUPPORT_PRODUCT_VERSION="7"
/etc/os-release:
/etc/redhat-release:CentOS Linux release 7.8.2003 (Core)
/etc/system-release:CentOS Linux release 7.8.2003 (Core)
```
##### Netdata version

  currently its removed. Installed yesterday using kickstrat script. so version will be Netdata v1.23.1 latest.
##### Component Name

kprobe events conflicts with kernel patching
##### Steps To Reproduce

1. Install cPanel on CentOS 7
2. Install KernelCare
3. Let kernelCare patch the kernel using cronjobs or run `/usr/bin/kcarectl --update` It will give an error.
Unable to apply patch (/var/cache/kcare/patches/none-1773540921d02e0d50f5433c8c3a4d5a02a14c33-5-edf/kpatch.edf.bin 5 255 CentOS Linux 3.10.0-1127.el7.x86_64, default, , , False)
More information from KernelCare Support:
http://prntscr.com/tbt6ih
4. uninstall netdata
5. try kernel patching and it fails.
6. remove kprobe events.
6. kernel patching successfully.
7. Server crashes to kernel panic and booting stuck with kprobe errors.
images: http://prntscr.com/tbrlly
http://prntscr.com/tbrprn
8. booted server to rescue mode
9. manual removal the netdata lines from the /sys/kernel/debug/tracing/kprobe_events file.
10. booted the system and centos kernel booted fine
##### Expected behavior

1) Uninstall should remove all the kprobe events from the system
2) Support for KernelCare patching on cPanel Servers.

### LINKED ISSUE #9522: ebpf plugin memory violation
Installed Netdata on a second web server today from script. Works fine except I can't get apache logs to show up. This is my latest error in /var/log/audit/audit.log

type=ANOM_ABEND msg=audit(1594422517.795:1589560): auid=4294967295 uid=994 gid=991 ses=4294967295 subj=system_u:system_r:unconfined_service_t:s0 pid=19998 comm="ebpf.plugin" reason="memory violation" sig=6

I had followed this tutorial and it seemed to bypass SELinux. I currently have setenforce 0 so its not caused by SELinux.
https://learn.netdata.cloud/docs/agent/collectors/ebpf.plugin#creation-of-bpf-policies

##### OS / Environment
/etc/centos-release:CentOS Linux release 7.7.1908 (Core)
/etc/os-release:NAME="CentOS Linux"
/etc/os-release:VERSION="7 (Core)"
/etc/os-release:ID="centos"
/etc/os-release:ID_LIKE="rhel fedora"
/etc/os-release:VERSION_ID="7"
/etc/os-release:PRETTY_NAME="CentOS Linux 7 (Core)"
/etc/os-release:ANSI_COLOR="0;31"
/etc/os-release:CPE_NAME="cpe:/o:centos:centos:7"
/etc/os-release:HOME_URL="https://www.centos.org/"
/etc/os-release:BUG_REPORT_URL="https://bugs.centos.org/"
/etc/os-release:
/etc/os-release:CENTOS_MANTISBT_PROJECT="CentOS-7"
/etc/os-release:CENTOS_MANTISBT_PROJECT_VERSION="7"
/etc/os-release:REDHAT_SUPPORT_PRODUCT="centos"
/etc/os-release:REDHAT_SUPPORT_PRODUCT_VERSION="7"
/etc/os-release:
/etc/redhat-release:CentOS Linux release 7.7.1908 (Core)
/etc/system-release:CentOS Linux release 7.7.1908 (Core)

##### Netdata version
netdata v1.23.1-62-nightly

### LINKED ISSUE #9529: ebpf.plugin segfault
On every machine we installed netdata we have many errors such as:

[Mon Jul 13 06:39:01 2020] EBPF PROCESS[20218]: segfault at 6410 ip 0000000000006410 sp 00007f3f77e3fb78 error 14 in ebpf.plugin (deleted)[55973a385000+10000]
[Mon Jul 13 06:39:08 2020] trace_kprobe: Event kprobes/r_netdata_vfs_write_0 doesn't exist.
[Mon Jul 13 06:39:08 2020] trace_kprobe: Event kprobes/r_netdata_tcp_sendmsg_0 doesn't exist.

All machines are either Ubuntu servers or Mint warkstations:
/etc/os-release:VERSION="18.04.4 LTS (Bionic Beaver)"

Netdata is v1.23.1-77-nightly

The above error doesn't seem to create any problem on the system itself however monitoring systems keep on alerting.

### PR DESCRIPTION
**Pull request by thiagoftsm**
##### Summary
Fixes #9424 
Fixes #9480 
Fixes #9522
Fixes #9529 

This PR fixes some errors that ebpf.plugin was stopping some times(memory leak), we decided to use the same approach that apps.plugin is using to clean its link list.
I am also converting the `error` call to `info` call for the users do not have the wrong impression that there is a bug when files related to kernel are not loaded.

##### Component Name
Plugin
##### Test Plan
There are two possible ways to test this PR:

###### Review kernel-collector fist

1 - Review the [PR](https://github.com/netdata/kernel-collector/pull/167), because it will force @thiagoftsm to create a new release that will simplify your life.
2 - Install this PR.
3 - Remove the file `/etc/netdata/ebpf.conf` or change the parameter `ebpf load mode` to `entry`.
4 - Start Netdata
5 - Stop Netdata
6 - Confirm that `/sys/kernel/debug/tracing/kprobe_events` was cleaned.
7 - Confirm tha you have `INFO` message instead `ERROR` messages associated to ebpf.plugin.
8 - Create the file `./edit-config ebpf.conf` or change the parameter `ebpf load mode` to `return`.
9 - Start Netdata again
10 - Stop Netdata.
11 - Repeat steps 6 and 7. 

###### Extracting files

1 - Download  [shared libraries](https://github.com/netdata/netdata/files/4875781/files.zip)
2 - Install this PR using `./netdata-installer.sh --dont-start-it`
3 - Extract the `shared libraries` inside `/usr/libexec/netdata/plugins.d`.
4 - Remove the file `/etc/netdata/ebpf.conf` or change the parameter `ebpf load mode` to `entry`.
5 - Start Netdata
6 - Stop Netdata
7 - Confirm that `/sys/kernel/debug/tracing/kprobe_events` was cleaned.
8 - Confirm tha you have `INFO` message instead `ERROR` messages associated to ebpf.plugin.
9 - Create the file `./edit-config ebpf.conf` or change the parameter `ebpf load mode` to `return`.
10 - Start Netdata again
11 - Stop Netdata.
12 - Repeat steps 7 and 8.

### CODE REVIEW

**Comment by thiagoftsm** [comment]
@prologic and @Ferroin I am changing the way we are stopping Netdata when we are using a `systemd` environment, because the user reported that on old kernels ( read CentOS 7.x) he had problem when he uninstall Netdata. I confirmed that this problem is real in this [link](https://books.google.com.br/books?id=ihTADwAAQBAJ&pg=PT212&lpg=PT212&dq=safe+way+to+clean+kprobe_events&source=bl&ots=bySO8CzEMo&sig=ACfU3U38VVVWE0T99v2WWdvTjSZNK5tC2Q&hl=en&sa=X&ved=2ahUKEwiIyr2W2bvqAhWbH7kGHd06CEkQ6AEwAHoECAoQAQ#v=onepage&q=safe%20way%20to%20clean%20kprobe_events&f=false) . Our script to clean probes only affects our probes different of the book, please, take a look to confirm I am not missing nothing.

**Comment by prologic** [comment]
@thiagoftsm Are the 31 build failures expected here? :)

**Comment by thiagoftsm** [comment]
> @thiagoftsm Are the 31 build failures expected here? :)

I inserted the wrong file inside `.gitignore` :/

**Comment by underhood** [comment]
Screenshot from 2020-07-13 19-02-17

Minor comment `ebpf.udp_functions` both sent and received dimensions are positive but in the chart below `ebpf.udp_bandwith` they oppose each other.

I think it should be consistent and I prefer the way that they are opposite.

Also, I would use the area chart for this kind of chart that has only 2 dimensions.

**Comment by thiagoftsm** [comment]
Hi @underhood ,

> Minor comment (i noticed during checking this PR but is not directly related) `ebpf.udp_functions` both sent and received dimensions are positive but in the chart below `ebpf.udp_bandwith` they oppose each other.

I understood your point, but these charts have different measures, while the second is plotting bandwidth, the first is counting number of calls for a specific function, I do not remember to see negative counts on Netdata (@ilyam8 do we have a pattern like this?)

**Comment by ilyam8** [comment]
>  do we have a pattern like this?

Check the dashboard - usually `rx` is positive, `tx` is negative (Network interfaces), same for `read`/`write` (Disks). I dont say it is good imo, just mentioning the pattern.

File: system/netdata.service.in [hunk 1]
```diff
@@ -17,7 +17,7 @@ ExecStartPre=/bin/mkdir -p @localstatedir_POST@/cache/netdata
 ExecStartPre=/bin/chown -R netdata:netdata @localstatedir_POST@/cache/netdata
 ExecStartPre=/bin/mkdir -p @localstatedir_POST@/run/netdata
 ExecStartPre=/bin/chown -R netdata:netdata @localstatedir_POST@/run/netdata
-#ExecStopPost=/bin/rm @localstatedir_POST@/run/netdata/netdata.pid
+ExecStopPost=/usr/libexec/netdata/plugins.d/reset_netdata_trace.sh
```
Comment (vlvkobal) [inline system/netdata.service.in:20]: `plugins.d` can be installed to a different directory.

File: .gitignore [hunk 2]
```diff
@@ -67,6 +67,7 @@ cgroup-network
 !cgroup-network/
 
 ebpf.plugin
+collectors/ebpf.plugin/reset_netdata_trace.sh
```
Comment (vlvkobal) [inline .gitignore:70]: You included the file to `.gitignore` but forgot to delete it from the repository.

Comment (thiagoftsm) [reply .gitignore:70]: I removed it, thank you to call attention for this.

File: system/netdata.service.in (hunk 1 above)
Comment (thiagoftsm) [reply system/netdata.service.in]: I added the missing prefix, thank you!

File: system/netdata.service.in [hunk 3]
```diff
@@ -17,7 +17,7 @@ ExecStartPre=/bin/mkdir -p @localstatedir_POST@/cache/netdata
 ExecStartPre=/bin/chown -R netdata:netdata @localstatedir_POST@/cache/netdata
 ExecStartPre=/bin/mkdir -p @localstatedir_POST@/run/netdata
 ExecStartPre=/bin/chown -R netdata:netdata @localstatedir_POST@/run/netdata
-#ExecStopPost=/bin/rm @localstatedir_POST@/run/netdata/netdata.pid
+ExecStopPost=@pluginsdir_POST@/usr/libexec/netdata/plugins.d/reset_netdata_trace.sh
```
Comment (vlvkobal) [inline system/netdata.service.in:20]:
```
 netdata.service: Failed at step EXEC spawning /usr/libexec/netdata/plugins.d/usr/libexec/netdata/plugins.d/reset_netdata_trace.sh: No such file or directory
```

Comment (thiagoftsm) [reply system/netdata.service.in]: Fixed with the last commit!

**Pull request merged**